src/haystack/encoding/json/encode.rs

```diff
@@ -54,8 +57,12 @@ impl Serialize for Ref {
         let mut map = serializer.serialize_map(Some(if self.dis.is_none() { 2 } else { 3 }))?;
         map.serialize_entry("_kind", "ref")?;
         map.serialize_entry("val", &self.value)?;
-        if self.dis.is_some() {
-            let dis = self.dis.clone();
+
+        let dis = get_ref_dis_from_factory(self.value.as_str(), self.dis.as_deref())
+            .map(|v| v.to_string())
```

Comment:
nitpick: Prefer `v.into_owned()`. `to_string` is a method from the `Display` trait, meant to print types as formatted strings. The result is the same, but `Cow::into_owned` provides better semantics since the objective is to create an owned value, not a formatted string. Besides, `to_string` calls the formatting logic, whereas `into_owned` simply clones the wrapped value if required.